File: src/app/lib/plc/parser.ts
import {
  Assign,
  AST,
  Call,
  CaseNode,
  CommentNode,
  Expr,
  ForNode,
  FunctionBlockDecl,
  FunctionDecl,
  IfNode,
  Program,
  VarDecl,
  WhileNode
} from "./ast";

import { Token } from "./lexer";

/**
 * =========================================================
 * Parser (Final IEC 61131-3 ST Parser)
 * =========================================================
 */
export function parser(tokens: Token[]): Program {

  const iRef = { i: 0 };

  const peek = () => tokens[iRef.i];
  const next = () => tokens[iRef.i++];

  const is = (t: any, v: string) => t && t.value === v;

  const PREC: Record<string, number> = {
    OR: 1,
    AND: 2,
    "==": 3,
    "<>": 3,
    "<": 3,
    ">": 3,
    "+": 4,
    "-": 4,
    "*": 5,
    "/": 5,
  };

  // =========================================================
  // Expression
  // =========================================================
  function parseExpr(minPrec = 0): Expr {

    function primary(): Expr {
      const t = next();
      if (!t) return { type: "num", value: 0 };

      // NOT unary
      if (t.type === "kw" && t.value === "NOT") {
        return {
          type: "unary",
          op: "not",
          value: parseExpr(6)
        };
      }

      if (t.type === "num") return { type: "num", value: t.value as number };
      if (t.type === "str") return { type: "str", value: t.value as string };

      if (t.type === "kw") {
        if (t.value === "TRUE") return { type: "bool", value: true };
        if (t.value === "FALSE") return { type: "bool", value: false };
      }

      if (t.type === "id") {
        const name = t.value as string;

        // function call
        if (peek()?.value === "(") {
          next();
          const args: Expr[] = [];

          while (peek() && peek().value !== ")") {
            args.push(parseExpr());
            if (peek()?.value === ",") next();
          }

          next();
          return { type: "call", name, args };
        }

        return { type: "var", name };
      }

      return { type: "num", value: 0 };
    }

    let left = primary();

    while (true) {
      const t = peek();
      if (!t) break;

      let op = t.value as string;

      // normalize AND / OR
      if (t.type === "kw") {
        if (op !== "AND" && op !== "OR") break;
      }

      const prec = PREC[op];
      if (prec === undefined || prec < minPrec) break;

      next();

      const right = parseExpr(prec + 1);

      left = {
        type: "bin",
        op: op,
        left,
        right
      } as any;
    }

    return left;
  }

  // =========================================================
  // Assign
  // =========================================================
  function parseAssign(): Assign {
    const left = next().value as string;
    next(); // :=

    return {
      type: "Assign",
      left,
      right: parseExpr()
    };
  }

  function parseCall(): Call | any {
    const name = next().value as string;
    next(); // (

    const args: any[] = [];

    while (peek() && peek().value !== ")") {

      // 🔥 命名参数 IN := xxx
      if (
        peek()?.type === "id" &&
        tokens[iRef.i + 1]?.value === ":="
      ) {
        const argName = next().value as string;
        next(); // :=

        const value = parseExpr();

        args.push({ name: argName, value });

      } else {
        // 普通函数参数
        args.push(parseExpr());
      }

      if (peek()?.value === ",") next();
    }

    next(); // )

    // 🔥 判断是否 FBCall（是否包含命名参数）
    const hasNamed = args.some(a => typeof a === "object" && a.name);

    if (hasNamed) {
      return {
        type: "FBCall",
        name,
        args
      };
    }

    return {
      type: "Call",
      name,
      args
    };
  }

  // =========================================================
  // Comment
  // =========================================================
  function parseComment(): CommentNode | undefined {
    const t = next();
    if (!t) return;

    if (t.type == "comment")
    return {
      type: "Comment",
      kind: t.kind,
      value: t.value
    };
    return undefined
  }

  // =========================================================
  // IF
  // =========================================================
  function parseIf(): IfNode {
    next();
    const cond = parseExpr();
    next(); // THEN

    const then: AST[] = [];
    const elseif: any[] = [];
    let elseBody: AST[] | undefined;

    while (peek() && !is(peek(), "END_IF")) {

      if (is(peek(), "ELSIF")) {
        next();
        const c = parseExpr();
        next();

        const body: AST[] = [];
        while (peek() && !["ELSIF", "ELSE", "END_IF"].includes(peek().value as string)) {
          const s = parseStmt();
          if (s) body.push(s);
        }

        elseif.push({ cond: c, body });
        continue;
      }

      if (is(peek(), "ELSE")) {
        next();
        elseBody = [];

        while (peek() && !is(peek(), "END_IF")) {
          const s = parseStmt();
          if (s) elseBody.push(s);
        }
        break;
      }

      const s = parseStmt();
      if (s) then.push(s);
    }

    next();

    return { type: "If", cond, then, elseif, else: elseBody };
  }

  // =========================================================
  // WHILE
  // =========================================================
  function parseWhile(): WhileNode {
    next();
    const cond = parseExpr();
    next(); // DO

    const body: AST[] = [];

    while (peek() && !is(peek(), "END_WHILE")) {
      const s = parseStmt();
      if (s) body.push(s);
    }

    next();

    return { type: "While", cond, body };
  }

  // =========================================================
  // FOR
  // =========================================================
  function parseFor(): ForNode {
    next();

    const v = next().value as string;
    next(); // :=

    const from = parseExpr();
    next(); // TO

    const to = parseExpr();

    let step: Expr | undefined;

    if (is(peek(), "BY")) {
      next();
      step = parseExpr();
    }

    next(); // DO

    const body: AST[] = [];

    while (peek() && !is(peek(), "END_FOR")) {
      const s = parseStmt();
      if (s) body.push(s);
    }

    next();

    return { type: "For", v, from, to, step, body };
  }

  // =========================================================
  // CASE
  // =========================================================
  function parseCase(): CaseNode {
    next();

    const expr = parseExpr();
    next(); // OF

    const branches: any[] = [];

    while (peek() && peek().value !== "END_CASE") {

      const value = parseExpr();

      if (peek()?.value === ":") next();

      const body: AST[] = [];

      while (peek()) {
        const t = peek();
        if (!t) break;

        if (t.value === "END_CASE") break;

        if (
          (t.type === "num" || t.type === "str" || t.type === "id") &&
          tokens[iRef.i + 1]?.value === ":"
        ) break;

        const s = parseStmt();
        if (s) body.push(s);
      }

      branches.push({ value, body });
    }

    next();

    return { type: "Case", expr, branches };
  }

  // =========================================================
  // VAR DECL
  // =========================================================
  function parseVarDecl(scope: any): VarDecl {
    next();

    const vars: any[] = [];

    while (peek() && peek().value !== "END_VAR") {

      const name = next().value as string;

      let dataType: string | undefined;
      let init;

      if (peek()?.value === ":") {
        next();
        dataType = next().value as string;
      }

      if (peek()?.value === ":=") {
        next();
        init = parseExpr();
      }

      if (peek()?.value === ";") next();

      vars.push({ name, dataType, init });
    }

    next();

    return { type: "VarDecl", scope, vars };
  }

  // =========================================================
  // FUNCTION
  // =========================================================
  function parseFunction(): FunctionDecl {
    next();

    const name = next().value as string;

    const body: AST[] = [];

    while (peek() && peek().value !== "END_FUNCTION") {
      const s = parseStmt();
      if (s) body.push(s);
    }

    next();

    return {
      type: "Function",
      name,
      params: [],
      body
    };
  }

  // =========================================================
  // FUNCTION BLOCK
  // =========================================================
  function parseFB(): FunctionBlockDecl {
    next();

    const name = next().value as string;

    const vars = {
      input: [],
      output: [],
      inout: [],
      local: [],
      temp: []
    };

    const body: AST[] = [];

    while (peek() && peek().value !== "END_FUNCTION_BLOCK") {
      const s = parseStmt();
      if (s) body.push(s);
    }

    next();

    return {
      type: "FunctionBlock",
      name,
      vars,
      body
    };
  }

  // =========================================================
  // DISPATCHER
  // =========================================================
  function parseStmt(): AST | undefined {

    const t = peek();
    if (!t) return;

    if (t.type === "comment") return parseComment();

    if (is(t, "IF")) return parseIf();
    if (is(t, "CASE")) return parseCase();
    if (is(t, "WHILE")) return parseWhile();
    if (is(t, "FOR")) return parseFor();

    if (is(t, "FUNCTION")) return parseFunction();
    if (is(t, "FUNCTION_BLOCK")) return parseFB();

    if (t.type === "kw" && (t.value as string).startsWith("VAR")) {
      return parseVarDecl(t.value);
    }

    if (t.type === "id" && tokens[iRef.i + 1]?.value === ":=") {
      return parseAssign();
    }

    if (t.type === "id" && tokens[iRef.i + 1]?.value === "(") {
      return parseCall();
    }

    next();
    return;
  }

  // =========================================================
  // PROGRAM
  // =========================================================
  const body: AST[] = [];

  while (iRef.i < tokens.length) {
    const s = parseStmt();
    if (s) body.push(s);
  }

  return { type: "Program", body };
}

File: plc/st/parser.go
package st

import (
	"fmt"
)

// =========================================================
// Parser
// =========================================================

type Parser struct {
	l         *Lexer
	curToken  Token
	peekToken Token

	inFunction bool
	fnName     string
}

func NewParser(l *Lexer) *Parser {
	p := &Parser{l: l}
	p.next()
	p.next()
	return p
}

func (p *Parser) next() {
	p.curToken = p.peekToken
	p.peekToken = p.l.NextToken()
}

func (p *Parser) expect(t TokenType) {
	if p.curToken.Type != t {
		panic(fmt.Sprintf("expected %v got %v (%s)",
			t, p.curToken.Type, p.curToken.Lit))
	}
	p.next()
}

// =========================================================
// PROGRAM
// =========================================================

func (p *Parser) ParseProgram() *Program {
	prog := &Program{}

	for p.curToken.Type != PROGRAM && p.curToken.Type != EOF {
		p.next()
	}

	p.expect(PROGRAM)
	prog.Name = p.curToken.Lit
	p.expect(IDENT)

	for p.curToken.Type != END_PROGRAM && p.curToken.Type != EOF {

		switch p.curToken.Type {

		case VAR, VAR_INPUT, VAR_OUTPUT, VAR_IN_OUT, VAR_GLOBAL:
			prog.Blocks = append(prog.Blocks, p.parseVarBlock())

		case FUNCTION_BLOCK:
			prog.Blocks = append(prog.Blocks, p.parseFunctionBlock())

		case FUNCTION:
			prog.Blocks = append(prog.Blocks, p.parseFunction())

		default:
			if p.curToken.Type == SEMI {
				p.next()
				continue
			}
			prog.Body = append(prog.Body, p.parseStatement())
		}
	}

	p.expect(END_PROGRAM)
	return prog
}

// =========================================================
// FUNCTION
// =========================================================

func (p *Parser) parseFunction() DeclBlock {
	fn := &Function{}

	p.expect(FUNCTION)
	fn.Name = p.curToken.Lit
	p.expect(IDENT)

	if p.curToken.Type == COLON {
		p.next()
		fn.ReturnType = p.parseType()
	}

	p.inFunction = true
	p.fnName = fn.Name

	for p.curToken.Type != END_FUNCTION && p.curToken.Type != EOF {

		if isVarBlock(p.curToken.Type) {
			p.parseVarBlock()
			continue
		}

		if p.curToken.Type == SEMI {
			p.next()
			continue
		}

		fn.Body = append(fn.Body, p.parseStatement())
	}

	p.inFunction = false
	p.expect(END_FUNCTION)
	return fn
}

// =========================================================
// FUNCTION_BLOCK
// =========================================================

func (p *Parser) parseFunctionBlock() DeclBlock {
	fb := &FunctionBlock{}

	p.expect(FUNCTION_BLOCK)
	fb.Name = p.curToken.Lit
	p.expect(IDENT)

	for p.curToken.Type != END_FUNCTION_BLOCK && p.curToken.Type != EOF {

		if isVarBlock(p.curToken.Type) {
			p.parseVarBlock()
			continue
		}

		if p.curToken.Type == SEMI {
			p.next()
			continue
		}

		fb.Body = append(fb.Body, p.parseStatement())
	}

	p.expect(END_FUNCTION_BLOCK)
	return fb
}

// =========================================================
// VAR BLOCK
// =========================================================

func isVarBlock(t TokenType) bool {
	return t == VAR ||
		t == VAR_INPUT ||
		t == VAR_OUTPUT ||
		t == VAR_IN_OUT ||
		t == VAR_GLOBAL
}

func (p *Parser) parseVarBlock() DeclBlock {
	v := &VarBlock{}

	v.Kind = p.curToken.Lit
	p.next()

	for p.curToken.Type != END_VAR && p.curToken.Type != EOF {
		v.Vars = append(v.Vars, p.parseVarDecl())
	}

	p.expect(END_VAR)
	return v
}

func (p *Parser) parseVarDecl() VarDecl {
	var d VarDecl

	d.Names = p.parseIdentList()

	p.expect(COLON)
	d.Type = p.parseType()

	if p.curToken.Type == ASSIGN {
		p.next()
		d.Init = p.parseExpression()
	}

	p.expect(SEMI)
	return d
}

func (p *Parser) parseIdentList() []string {
	var ids []string

	ids = append(ids, p.curToken.Lit)
	p.expect(IDENT)

	for p.curToken.Type == COMMA {
		p.next()
		ids = append(ids, p.curToken.Lit)
		p.expect(IDENT)
	}

	return ids
}

// =========================================================
// TYPE
// =========================================================

func (p *Parser) parseType() Type {
	if p.curToken.Type == IDENT {
		t := &BasicType{Name: p.curToken.Lit}
		p.next()
		return t
	}
	panic("unknown type")
}

// =========================================================
// STATEMENTS
// =========================================================

func (p *Parser) parseStatement() Stmt {
	switch p.curToken.Type {

	case IF:
		return p.parseIf()

	case FOR:
		return p.parseFor()

	case WHILE:
		return p.parseWhile()

	case RETURN:
		return p.parseReturn()

	case CASE:
		return p.parseCase()

	case IDENT:
		return p.parseAssignOrCall()

	case SEMI:
		p.next()
		return nil
	}

	panic(fmt.Sprintf("unknown stmt %v (%s)",
		p.curToken.Type, p.curToken.Lit))
}

// =========================================================
// IF
// =========================================================

func (p *Parser) parseIf() *IfStmt {
	stmt := &IfStmt{}

	p.expect(IF)
	stmt.Cond = p.parseExpression()
	p.expect(THEN)

	stmt.Then = p.parseBlock()

	for p.curToken.Type == ELSIF {
		p.next()
		cond := p.parseExpression()
		p.expect(THEN)

		stmt.ElseIf = append(stmt.ElseIf, ElseIfBranch{
			Cond: cond,
			Body: p.parseBlock(),
		})
	}

	if p.curToken.Type == ELSE {
		p.next()
		stmt.Else = p.parseBlock()
	}

	p.expect(END_IF)
	return stmt
}

// =========================================================
// FOR
// =========================================================

func (p *Parser) parseFor() *ForStmt {
	stmt := &ForStmt{}

	p.expect(FOR)
	stmt.Var = p.curToken.Lit
	p.expect(IDENT)

	p.expect(ASSIGN)
	stmt.From = p.parseExpression()

	p.expect(TO)
	stmt.To = p.parseExpression()

	if p.curToken.Type == BY {
		p.next()
		stmt.By = p.parseExpression()
	}

	p.expect(DO)
	stmt.Body = p.parseBlock()

	p.expect(END_FOR)
	return stmt
}

// =========================================================
// WHILE
// =========================================================

func (p *Parser) parseWhile() *WhileStmt {
	stmt := &WhileStmt{}

	p.expect(WHILE)
	stmt.Cond = p.parseExpression()

	p.expect(DO)
	stmt.Body = p.parseBlock()

	p.expect(END_WHILE)
	return stmt
}

// =========================================================
// RETURN
// =========================================================

func (p *Parser) parseReturn() *ReturnStmt {
	r := &ReturnStmt{}

	p.expect(RETURN)

	if p.curToken.Type != SEMI {
		r.Value = p.parseExpression()
	}

	p.expect(SEMI)
	return r
}

// =========================================================
// CASE (FINAL MERGED VERSION)
// =========================================================

func (p *Parser) parseCase() *CaseStmt {
	c := &CaseStmt{}

	p.expect(CASE)
	c.Expr = p.parseExpression()
	p.expect(OF)

	var branch CaseBranch
	var isElse bool

	for p.curToken.Type != END_CASE && p.curToken.Type != EOF {

		if p.curToken.Type == ELSE {
			p.next()
			isElse = true
			continue
		}

		//识别Label
		if p.curToken.Type == NUMBER {
			//保存上一个
			if len(branch.Values) > 0 {
				c.Branches = append(c.Branches, branch)
			}

			branch = CaseBranch{}
			branch.Values = append(branch.Values, p.parseExpression())

			for p.curToken.Type == COMMA {
				p.next()
				branch.Values = append(branch.Values, p.parseExpression())
			}

			p.expect(COLON)
			continue
		}

		//逐行识别
		if isElse {
			c.Else = append(c.Else, p.parseStatement())
		} else {
			branch.Body = append(branch.Body, p.parseStatement())
		}
	}

	//保存上一个
	if len(branch.Values) > 0 {
		c.Branches = append(c.Branches, branch)
	}

	p.expect(END_CASE)
	return c
}

// =========================================================
// LVALUE
// =========================================================

func (p *Parser) parseLValue() Expr {
	name := p.curToken.Lit
	p.expect(IDENT)

	parts := []string{name}

	for p.curToken.Type == DOT {
		p.next()
		parts = append(parts, p.curToken.Lit)
		p.expect(IDENT)
	}

	if p.curToken.Type == LPAREN {
		call := &CallExpr{Name: parts[len(parts)-1]}
		call.Args = p.parseArgs()
		return call
	}

	return &VarExpr{
		Path: parts,
	}
}

// =========================================================
// ARGS
// =========================================================

func (p *Parser) parseArgs() []Param {
	var args []Param

	p.expect(LPAREN)

	for p.curToken.Type != RPAREN {

		arg := Param{}
		arg.Name = p.curToken.Lit
		p.expect(IDENT)

		p.expect(ASSIGN)
		arg.Value = p.parseExpression()

		args = append(args, arg)

		if p.curToken.Type == COMMA {
			p.next()
		}
	}

	p.expect(RPAREN)
	return args
}

// =========================================================
// ASSIGN / CALL
// =========================================================

func (p *Parser) parseAssignOrCall() Stmt {
	expr := p.parseLValue()

	if p.curToken.Type == ASSIGN {
		p.next()
		right := p.parseExpression()
		p.expect(SEMI)

		return &AssignStmt{
			Left:  expr,
			Right: right,
		}
	}

	if call, ok := expr.(*CallExpr); ok {
		p.expect(SEMI)
		return &CallStmt{Call: call}
	}

	panic("invalid statement")
}

// =========================================================
// BLOCK
// =========================================================

func (p *Parser) parseBlock() []Stmt {
	var stmts []Stmt

	for !isBlockEnd(p.curToken.Type) {

		if p.curToken.Type == SEMI {
			p.next()
			continue
		}

		stmts = append(stmts, p.parseStatement())
	}

	return stmts
}

func isBlockEnd(t TokenType) bool {
	return t == END_IF ||
		t == END_FOR ||
		t == END_WHILE ||
		t == END_PROGRAM ||
		t == END_FUNCTION ||
		t == END_FUNCTION_BLOCK ||
		t == END_CASE ||
		t == ELSE ||
		t == ELSIF ||
		t == EOF
}

// =========================================================
// EXPRESSIONS (PRATT)
// =========================================================

func (p *Parser) parseExpression() Expr {
	return p.parseBinary(0)
}

var precedence = map[TokenType]int{
	OR:    1,
	AND:   2,
	EQ:    3,
	NEQ:   3,
	LT:    4,
	LTE:   4,
	GT:    4,
	GTE:   4,
	PLUS:  5,
	MINUS: 5,
	MUL:   6,
	DIV:   6,
}

func (p *Parser) parseBinary(min int) Expr {
	left := p.parsePrimary()

	for {
		prec, ok := precedence[p.curToken.Type]
		if !ok || prec < min {
			break
		}

		op := p.curToken.Type
		p.next()

		right := p.parseBinary(prec + 1)

		left = &BinaryExpr{
			Left:  left,
			Op:    op.String(),
			Right: right,
		}
	}

	return left
}

func (p *Parser) parsePrimary() Expr {
	switch p.curToken.Type {

	case NUMBER:
		v := atof(p.curToken.Lit)
		p.next()
		return &NumberLit{Value: v}

	case BOOL:
		v := p.curToken.Lit == "TRUE"
		p.next()
		return &BoolLit{Value: v}

	case STRING:
		v := p.curToken.Lit
		p.next()
		return &StringLit{Value: v}

	case IDENT:
		name := p.curToken.Lit
		p.next()
		return &VarExpr{Path: []string{name}}

	case LPAREN:
		p.next()
		e := p.parseExpression()
		p.expect(RPAREN)
		return e
	}

	panic(fmt.Sprintf("bad expr %v (%s)",
		p.curToken.Type, p.curToken.Lit))
}

// =========================================================
// HELPERS
// =========================================================

func atof(s string) float64 {
	var f float64
	fmt.Sscanf(s, "%f", &f)
	return f
}
